mpc/src/honeybadger/batch_recon/batch_recon.rs

```diff
@@ -1,213 +1,417 @@
+                let y_j = F::deserialize_compressed(msg.payload.as_slice())
+                    .map_err(|e| BatchReconError::ArkDeserialization(e))?;
+
+                // Store the received revealed `y_j` value if it's from a new sender.
+                if !self.reveals_received.iter().any(|s| s.id == sender_id) {
+                    self.reveals_received
+                        .push(RobustShamirShare::new(y_j, sender_id, self.t));
                 }
+                // Check if we have enough revealed `y_j` values and haven't already reconstructed the secrets.
                 if self.reveals_received.len() >= 2 * self.t + 1 && self.secrets.is_none() {
                     info!(
                         self_id = self.id,
                         "Enough Reveals collected, interpolating secrets"
                     );
-
-                    match robust_interpolate(self.n, self.t, self.reveals_received.clone()) {
-                        Some((poly, _)) => {
-                            let mut result = poly.coeffs;
+                    // Attempt to interpolate the polynomial whose coefficients are the original secrets.
+                    match RobustShamirShare::recover_secret(&self.reveals_received.clone()) {
+                        Ok((poly, _)) => {
+                            let mut result = poly;
+                            // Resize the coefficient vector to `t + 1` to get all secrets.
                             result.resize(self.t + 1, F::zero());
                             self.secrets = Some(result);
                             info!(self_id = self.id, "Secrets successfully reconstructed");
                         }
-                        None => {
+                        Err(e) => {
                             warn!(
-                                self_id = self.id,
-                                "Final secrets interpolation failed — possibly too many incorrect y_j"
+                                self_id = self.id, error = ?e,
+                                "Final secrets interpolation failed "
                             );
+                            return Err(BatchReconError::Inner(e));
                         }
                     }
                 }
+                Ok(())
             }
         }
     }
+    // pub async fn process<N: Network>(
+    //     &mut self,
+    //     raw_msg: Vec<u8>,
+    //     net: &Arc<N>,
+    // ) -> Result<(), BatchReconError> {
+    //     let msg: BatchReconMsg =
+    //         bincode::deserialize(&raw_msg).map_err(|e| BatchReconError::SerializationError(e))?;
+
+    //     self.batch_recon_handler(msg, net).await?;
+    //     Ok(())
+    // }
+}
+
+/// Creates a Vandermonde matrix `V` of size `n x (t+1)`.
+/// Each row `j` contains powers of `domain.element(j)`: `[1, alpha_j, alpha_j^2, ..., alpha_j^t]`.
+pub fn make_vandermonde<F: FftField>(n: usize, t: usize) -> Result<Vec<Vec<F>>, InterpolateError> {
```

Comment:
`make_vandermonde` and `apply_vandermonde` are general and not just applicable to batch reconstruction. We should probably move these functions to the higher level `mod.rs` so that it can be used for other asynchronous protocols in the future